arangod/IResearch/IResearchLink.cpp

```diff
@@ -915,30 +1088,88 @@ Result IResearchLink::commitUnsafeImpl(bool wait, CommitResult* code) {
 
       commitLock.lock();
     }
-
-    auto const lastCommittedTick = _lastCommittedTick;
-
+    _commitStageOne = true;
+    auto const lastCommittedTick = _lastCommittedTickStageOne;
+
+#if ARANGODB_ENABLE_MAINTAINER_MODE && ARANGODB_ENABLE_FAILURE_TESTS
+    TRI_IF_FAILURE("ArangoSearch::ThreeTransactionsMisorder") {
+      std::unique_lock<std::mutex> sync(_t3Failureync);
+      while (_t3NumFlushRegistered < 2) {
+        sync.unlock();
+        std::this_thread::sleep_for(100ms);
+        sync.lock();
+      }
+      _t3CommitSignal = true;
+      LOG_TOPIC("4cb66", DEBUG, TOPIC) << "Commit started";
+    }
+#endif
     try {
       // _lastCommittedTick is being updated in '_before_commit'
       *code = _dataStore._writer->commit() ? CommitResult::DONE
                                            : CommitResult::NO_CHANGES;
     } catch (...) {
       // restore last committed tick in case of any error
-      _lastCommittedTick = lastCommittedTick;
+      _lastCommittedTickStageOne = lastCommittedTick;
```

Comment:
do we need to reset `_commitStageOne` here?

Reply:
No, it would be resetted on the beginning on next commit/truncate